arch/riscv/core/pmp.c

```diff
@@ -317,3 +322,254 @@ void z_riscv_pmp_stackguard_enable(struct k_thread *thread)
 }
 
 #endif /* CONFIG_PMP_STACK_GUARD */
+
+#ifdef CONFIG_USERSPACE
+
+/**
+ * @brief Initialize the usermode portion of the PMP configuration.
+ *
+ * This is called once during new thread creation.
+ */
+void z_riscv_pmp_usermode_init(struct k_thread *thread)
+{
+	/* Only indicate that the usermode PMP is not prepared yet */
+	thread->arch.pmp_u_end_index = 0;
+}
+
+/**
+ * @brief Prepare the PMP usermode content for given thread.
+ *
+ * This is called once before making the transition to usermode.
+ */
+void z_riscv_pmp_usermode_prepare(struct k_thread *thread)
+{
+	unsigned int index = global_pmp_end_index;
+
+	/* Retrieve pmpcfg0 partial content from global entries */
+	thread->arch.pmp_u_cfg[0] = global_pmp_cfg[0];
+
+#if !defined(CONFIG_SMP)
+	/* Map the is_user_mode variable */
+	extern uint32_t is_user_mode;
+
+	set_pmp_entry(&index, PMP_R,
+		      (uintptr_t) &is_user_mode, sizeof(is_user_mode),
+		      thread->arch.pmp_u_addr, thread->arch.pmp_u_cfg,
+		      ARRAY_SIZE(thread->arch.pmp_u_addr));
+#endif
+
+	/* Map the usermode stack */
+	set_pmp_entry(&index, PMP_R | PMP_W,
+		      thread->stack_info.start, thread->stack_info.size,
+		      thread->arch.pmp_u_addr, thread->arch.pmp_u_cfg,
+		      ARRAY_SIZE(thread->arch.pmp_u_addr));
+
+	thread->arch.pmp_u_domain_offset = index;
+	thread->arch.pmp_u_end_index = index;
+	thread->arch.pmp_u_update_nr = 0;
+}
+
+/**
+ * @brief Convert partition information into PMP entries
+ */
+static void resync_pmp_domain(struct k_thread *thread,
+			      struct k_mem_domain *domain)
+{
+	unsigned int index = thread->arch.pmp_u_domain_offset;
+	int p_idx, remaining_partitions;
+	k_spinlock_key_t key = k_spin_lock(&z_mem_domain_lock);
+
+	remaining_partitions = domain->num_partitions;
+	for (p_idx = 0; remaining_partitions > 0; p_idx++) {
+		struct k_mem_partition *part = &domain->partitions[p_idx];
+
+		if (part->size == 0) {
+			/* skip empty partition */
+			continue;
+		}
+		remaining_partitions--;
+
+		if (part->size < 4) {
+			/*
+			 * 4 bytes is the minimum we can map. If the size
+			 * is too small and assertions are disabled then
+			 * we skip those as well for security reasons.
+			 */
+			__ASSERT(false, "non-empty partition too small");
+			continue;
+		}
+
+		set_pmp_entry(&index, part->attr.pmp_attr,
+			      part->start, part->size,
+			      thread->arch.pmp_u_addr,
+			      thread->arch.pmp_u_cfg,
+			      ARRAY_SIZE(thread->arch.pmp_u_addr));
+		__ASSERT(index <= ARRAY_SIZE(thread->arch.pmp_u_addr),
```

Comment:
If assertions are disabled, the requested memory domain won't be honored and thread will die due missing permissions, right? Isn't it at least worth an error message here, to aid debugging?

Reply:
Fair enough.